dali/python/nvidia/dali/worker.py

```diff
@@ -0,0 +1,257 @@
+            if processed_task is None:
+                self.ready_queue.insert(0, None)
+            else:
+                # assert len(ready_queue) < prefetch_queue_depths[scheduled.context_i], "Worker queue size exceeded."
+                self.ready_queue.append(processed_task)
+            self.ready_cv.notify()
+
+    def dispatcher_thread(self):
+        """Receives batches produced in the main thread and dispatches them to the parent process.
+        It is intended to be run in a separate thread because both callback and dispatcher may
+        wait on IO operations a lot and in that case Python threads provide some performance gain.
+        """
+        try:
+            while True:
+                message = self._wait_for_processed()
+                if message is None:
+                    break
+                self._send(message)
+        finally:
+            # In case of error, we don't know when exactly we were interrupted and the main process
+            # may be waiting for differant message that we would try to send. Close the communication
+            # to indicate an error for main process and allow the exception to propagate
+            # in the worker process.
+            self._shutdown()
+
+    def _wait_for_processed(self):
+        with self.ready_cv:
+            while len(self.ready_queue) == 0:
+                self.ready_cv.wait()
+            message = self.ready_queue.pop(0)
+        return message
+
+    def _send(self, processed_tasks: _ProcessedTasks):
+        """Send the processed task back to the main process"""
+        if processed_tasks.is_failed():  # one of the tasks failed
+            completed_tasks = CompletedTasks.failed(self.worker_id, processed_tasks)
+            self.res_pipe.send(completed_tasks)
+            return
+        serialized_batch = write_batch(processed_tasks.mem_chunk, processed_tasks.data_batch)
+        completed_tasks = CompletedTasks.done(self.worker_id, processed_tasks, serialized_batch)
+        self.res_pipe.send(completed_tasks)
+        # send shared memory handle for underlaying shared memory chunk
+        # if it hasn't been sent ever before
+        mem_chunk_id = serialized_batch.mem_chunk_id
+        if mem_chunk_id not in self.handle_sent:
+            self.handle_sent.add(mem_chunk_id)
+            reduction.send_handle(self.sock, processed_tasks.mem_chunk.shm_chunk.handle, os.getppid())
+
+    def _shutdown(self):
+        """Force to close all communication channels (sockets and pipes) to unlock main process
+        in case of error, when it may be waiting for messages that we can't deliver or the
+        state of protocol is mismatched"""
+        self.res_pipe.close()
+        self.sock.shutdown(socket.SHUT_RDWR)
+        self.sock.close()
+
+class TaskReceiver:
+    def __init__(self, task_pipe):
+        self.task_pipe = task_pipe
+        self.tasks_cv = threading.Condition()
+        self.tasks_queue = []
+
+    def get_task(self):
```

Comment:
It seems that get_task and _insert_task matches what dispatch and _wait_for_processed do. It could be extracted to a common class. Just an idea.

Reply:
I think it's easier to just repeat here at this point. Otherwise we would add more code with wrapper and how to use it I feel. It's not like we want this to look like CRTP.